src/equations/passive_tracers.jl

```diff
@@ -0,0 +1,222 @@
+
+    u_tracers = u_flow[1] * tracers
+    return SVector(u_flow..., u_tracers...)
+end
+
+# Calculate the number of variables for the flow equations
+@inline function nvariables_flow(tracer_equations::PassiveTracerEquations)
+    return nvariables(tracer_equations.flow_equations)
+end
+
+# Obtain the flow variables from the conservative variables. The tracers are not included.
+@inline function flow_variables(u, tracer_equations::PassiveTracerEquations)
+    n_flow_variables = nvariables_flow(tracer_equations)
+    return SVector((u[i] for i in 1:n_flow_variables)...)
+end
+
+# Obtain the tracers which advect by the flow velocity by dividing the respective conservative
+# variable by the density.
+function tracers(u, tracer_equations::PassiveTracerEquations)
+    n_flow_variables = nvariables_flow(tracer_equations)
+    n_tracers = ntracers(tracer_equations)
+
+    rho = density(u, tracer_equations)
+    return SVector((u[n_flow_variables + i] / rho for i in 1:n_tracers)...)
+end
+
+# Primitives for the flow equations and tracers. For a tracer, the primitive variable is obtained
+# by dividing by density.
+@inline function cons2prim(u, tracer_equations::PassiveTracerEquations)
+    return SVector(cons2prim(flow_variables(u, tracer_equations),
+                             tracer_equations.flow_equations)...,
+                   tracers(u, tracer_equations)...)
+end
+
+# Conservative variables for the flow equations and tracers. For tracers, the conservative variable
+# is obtained by multiplying by the density
+@inline function prim2cons(u, tracer_equations::PassiveTracerEquations)
+    @unpack flow_equations = tracer_equations
+    u_flow = flow_variables(u, tracer_equations)
+
+    n_flow_variables = nvariables_flow(tracer_equations)
+    n_tracers = ntracers(tracer_equations)
+
+    rho = density(u, tracer_equations)
+    return SVector(prim2cons(u_flow, flow_equations)...,
+                   (rho * u[n_flow_variables + i] for i in 1:n_tracers)...)
+end
+
+# Entropy for tracers is the L2 norm of the tracers
+@inline function entropy(cons, tracer_equations::PassiveTracerEquations)
+    flow_entropy = entropy(flow_variables(cons, tracer_equations),
+                           tracer_equations.flow_equations)
+    tracer_entropy = density(cons, tracer_equations) *
+                     sum(abs2, tracers(cons, tracer_equations))
+    return flow_entropy + tracer_entropy
+end
+
+# Convert conservative variables to entropy
+@inline function cons2entropy(u, tracer_equations::PassiveTracerEquations)
+    @unpack flow_equations = tracer_equations
+    flow_entropy = cons2entropy(flow_variables(u, tracer_equations), flow_equations)
+    tracers_ = tracers(u, tracer_equations)
+
+    variables = SVector(flow_entropy[1] - sum(tracers_ .^ 2),
+                        (flow_entropy[i] for i in 2:nvariables(flow_equations))...,
+                        2.0f0 * tracers_...)
```

Comment:
```suggestion
                        2 * tracers_...)
```

should get correctly promoted.

Additionally, maybe add a comment why the factor 2 appears?